emcc.py

```diff
@@ -2154,38 +2184,38 @@ def get_clang_command_asm(src_file):
       return get_compiler(use_cxx(src_file)) + get_clang_flags() + compile_args + [src_file]
 
     # preprocessor-only (-E) support
-    if has_dash_E or '-M' in newargs or '-MM' in newargs or '-fsyntax-only' in newargs:
+    if state.preprocess_only:
```

Comment:
strictly speaking `-fsyntax-only` includes preprocessing and parsing, so it's not preprocess_only. Actually I guess the `-M` flags are similar (they preprocess and output something other than a compiled file).
I guess my question is, why do we care about whether the output is compiled or not. e.g. whether the output is `.i` or `.o`?

Reply:
I'm not totally sure, but maybe it can be unified as a followup.

One difference looks like in the handling of output files.  If you don't specify and output filename then `-c` will create N different output files, one more input.  But I think that `-E`, `-M` act in a similar way which is to output everything concatinated to stdout.  (-fsyntax-only doesn't output anything, but still we don't need to worry about creating N different output filenames).

Reply:
You are correct that `-fsyntax-only` does more than preprocessing... I guess ideally we could have another mode for syntax only.   I'd rather live this this as is for now and correct that oversight in the followup which introduces the single global mode enumeration instead.

Reply:
Yeah, that makes sense, it looks like all these conditionals are more about how the output files are handled. But thinking about that in a followup seems fine, i just happened to notice it.